[serve][deploy refactor][3/X] Move deployment state manager into application state manager

zcin opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

The main goal is to adopt a proper reconciler pattern for the application state manager.
- This requires moving deployment state manager inside application state manager.
  - Previously, the controller would take care of telling the deployment state manager to deploy or delete applications. 
  - Now, 2.0 deployments are deployed through application state manager
  - 1.0 code path is: `controller.deploy()` -> `deployment_state_manager.deploy()` on a duplicated code path.
- When `serve.run` is run or a new config is deployed, this sets the application's target state.
- In the application state's `update()`, reconcile the target state. Repeatedly:
  - Deploy all deployments listed in the application's target state
  - Delete all (past) deployments that aren't part of the application's current target state
  - Update application status based on deployment statuses
- Checkpoint deployment infos per application. This is now necessary because with a reconciler pattern, deployments won't be checkpointed in the deployment state manager after the call to deploy a config returns; it takes the next iteration of `update()` for it to be checkpointed in the deployment state manager.
    - This is now redundant with the checkpointing in deployment state manager. We might be able to merge them somehow in the future.

Tests in `test_application_state`:
* Test different flows in application status:
  * Basic flow: `DEPLOYING -> RUNNING -> DELETING`
  * Application stuck on deploy failed: `DEPLOYING -> DEPLOY_FAILED -> (redeploy) -> DEPLOYING -> RUNNING`
  * Application recovers from deploy failed: `DEPLOYING -> DEPLOY_FAILED -> (recovers) -> RUNNING`
  * Application becomes unhealthy after successfully deploying: `DEPLOYING -> RUNNING -> UNHEALTHY -> RUNNING`
* Test deploy through config, which requires waiting for the deploy task to finish
  * If deploy task succeeds and deployments become healthy, application should transition to `RUNNING`
  * If deploy task fails, application should transition to `DEPLOY_FAILED`
* Redeploy same application with different deployments
* Test recovery
  * Test controller crashes while the application is running normally `DEPLOYING -> RUNNING -> (controller crash) -> DEPLOYING -> RUNNING`
  * Test controller crashes while the application is deploying `DEPLOYING -> (controller crash) -> DEPLOYING -> RUNNING`. In particular, the deployment versions should be correctly reconciled.

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

edoakes commented on python/ray/serve/_private/application_state.py:
```diff
@@ -109,8 +147,21 @@ def deploy(self, deployment_params: List[Dict]) -> List[str]:
                 "path in your application to avoid this issue."
             )
 
+        # Delete old deployments
+        self.deployments_to_delete = set(self.deployments).difference(
+            {params["name"] for params in deployment_params}
+        )
+        for deployment_name in self.deployments_to_delete:
+            self.delete_deployment(deployment_name)
+        self.deployment_params = deployment_params
+
+        # Deploy new deployments
+        updating = []
+        for deployment in deployment_params:
+            updating.append(self.deploy_deployment(**deployment))
```
I think for correctness this logic should happen in `update`. As written this could lead to an inconsistent internal state (e.g., if `deploy_deployment` failed, some deployments would be deleted and some would not be updated).

zcin commented on python/ray/serve/api.py:
```diff
@@ -536,10 +531,10 @@ def run(
     )
 
     if ingress is not None:
+        client._wait_for_deployment_created(ingress.name)
```
With the migration to a reconciler patter in application state manager, there is no guarantee that deployments are created after `controller.deploy_application()` returns; it will take an iteration of `update()` to create those deployments. So before we return a handle to the ingress deployment, we need to wait for the ingress deployment to be created.

sihanwang41 replied:
amazing catch

zcin replied:
@edoakes Yeah, if `_blocking=True` this isn't necessary because `deploy_application` will wait for the application to become running. It is only to fix the case when `_blocking=False`.

zcin commented:
@edoakes Have yet to add any new tests, but ready for first round of review!

sihanwang41 commented on python/ray/serve/_private/application_state.py:
```diff
@@ -85,35 +109,71 @@ def deploy_obj_ref(self) -> Optional[ObjectRef]:
+        deleting: bool = False,
+    ):
+        if deleting:
+            target_state = ApplicationTargetState(dict(), True)
```
what is the diff between dict() and None in `ApplicationTargetState` ?
If no, I feel we can set to `None` to avoid any confusion.

zcin replied:
`None` is for when a new config has been deployed and we're waiting on `deploy_obj_ref`. In this period we don't perform any reconciliation. I set target state to an empty dictionary so that in the reconciliation loop, all of the deployments are removed from the application. LMK if that's confusing.

sihanwang41 replied:
oh my question is do we need to have separate attribute value (None and dict()) to indicate different scenario inside the `ApplicationTargetState` ?

zcin replied:
The problem with only using one or the other is that when a new config is deployed, I want to stop reconciliation, which is different from setting to empty dictionary because that will remove all the deployments when a new config is deployed.

sihanwang41 commented on python/ray/serve/_private/application_state.py:
```diff
@@ -141,27 +201,28 @@ def deploy(self, deployment_params: List[Dict]) -> List[str]:
+        application state stops the reconciliation loop and waits on the task to finish
+        and call apply_deployments_args().
+        """
         self._deploy_obj_ref = deploy_obj_ref
```
We should cancel the previous obj ref in the following scenario
- delete application.
- redeploy same application.

sihanwang41 commented on python/ray/serve/_private/application_state.py:
```diff
@@ -85,35 +109,71 @@ def deploy_obj_ref(self) -> Optional[ObjectRef]:
+
+        return updating
+
+    def apply_deployment_args(self, deployment_params: List[Dict]) -> None:
```
This function will not work if the application under the deleting state. i think we can failed it loudly.

shrekris-anyscale commented on python/ray/serve/_private/application_state.py:
```diff
@@ -435,3 +623,25 @@ def update(self):
+
+        self._kv_store.put(
+            CHECKPOINT_KEY,
+            cloudpickle.dumps(application_state_info),
```
Do you need to use `cloudpickle` to dump and load `application_state_info`? If you can use `pickle` instead, I'd recommend doing that instead. It's faster, and it prevents code from being serialized.

shrekris-anyscale commented on python/ray/serve/_private/application_state.py:
```diff
@@ -1,24 +1,41 @@
+
+@dataclass
+class ApplicationTargetState:
+    # Map of deployment names to deployment infos
```
```suggestion
    """Maps deployment names to deployment infos."""
```

shrekris-anyscale commented on python/ray/serve/_private/application_state.py:
```diff
@@ -84,37 +117,94 @@ def deploy_obj_ref(self) -> Optional[ObjectRef]:
+        )
+
+        if deployment_info.route_prefix is not None:
+            assert deployment_info.route_prefix.startswith("/")
```
Should we `assert` this before calling [`self._deployment_state_manager.deploy(...)`](https://github.com/ray-project/ray/pull/34328/files#diff-31975e75fad7092f05d8d5759dfeb90d8b454fc4193e4970b55914dbb87ad968R180-R182)?

shrekris-anyscale commented on python/ray/serve/_private/application_state.py:
```diff
@@ -141,127 +231,191 @@ def deploy(self, deployment_params: List[Dict]) -> List[str]:
+        # have info on what the target list of deployments is, so we
+        # can't check on deployment statuses
+        if self._target_state.deployment_infos is None:
+            return
```
Should this return a boolean? If not, could you update the return type and the docstring to indicate that the return value is `Optional`?

zcin replied:
Yeah, this should return false - let me update

shrekris-anyscale commented on python/ray/serve/_private/application_state.py:
```diff
@@ -141,127 +231,191 @@ def deploy(self, deployment_params: List[Dict]) -> List[str]:
+        """Reconcile target deployments in application target state.
+
+        Ensure each deployment is running on up-to-date info, and
+        remove outdated deployments from application
```
```suggestion
        remove outdated deployments from the application.
```

shrekris-anyscale reviewed (commented):
Nice work so far! I took a brief look at the code. I haven't reviewed any of the tests.

zcin commented on dashboard/client/src/type/serve.ts:
```diff
@@ -5,6 +5,7 @@ export enum ServeApplicationStatus {
   RUNNING = "RUNNING",
   DEPLOY_FAILED = "DEPLOY_FAILED",
   DELETING = "DELETING",
+  UNHEALTHY = "UNHEALTHY",
```
Yeah, I will add the unhealthy state to Anyscale services and Kuberay

shrekris-anyscale reviewed (approved):
The change looks good overall. My only pending concern is [whether we need Cloudpickle](https://github.com/ray-project/ray/pull/34328#discussion_r1227236160). I haven't had a chance to look at the unit tests. Let me know if you need me to review those as well.

sihanwang41 commented on python/ray/serve/_private/application_state.py:
```diff
@@ -141,127 +237,191 @@ def deploy(self, deployment_params: List[Dict]) -> List[str]:
+        # If we're waiting on the build app task to finish, we don't
+        # have info on what the target list of deployments is, so we
+        # can't check on deployment statuses
+        if self._target_state.deployment_infos is None:
```
is there is a case when object_ref = None too?

nit: 
if self._deploy_obj_ref:
    return Fasle

zcin replied:
I reset the deploy_obj_ref to None once we fetch the result from the deploy task, so if deploy_obj_ref=None, it just means that the task has finished.

zcin commented:
> @zcin , can you help to add a test redeploy the same app while app is under deletion in application state test? (ignore this message if I miss it)

@sihanwang41 Ah this test exists in `test_application_state`. The behavior can be improved (tracked in https://github.com/anyscale/product/issues/18753), but planning to make the improvements separately to avoid overloading this PR.

sihanwang41 commented on python/ray/serve/_private/client.py:
```diff
@@ -219,6 +220,53 @@ def _wait_for_deployment_deleted(self, name: str, timeout_s: int = 60):
+                f"Deployment {name} did not become HEALTHY after {timeout_s}s."
+            )
+
+    def _wait_for_application_running(self, name: str, timeout_s: int = -1):
```
raise error if app name not exsited.

sihanwang41 commented on python/ray/serve/_private/client.py:
```diff
@@ -279,38 +326,18 @@ def deploy_application(
-        if remove_past_deployments:
-            # clean up the old deployments
-            new_deployments_names = set()
+        if _blocking:
```
duplicated `if _blocking:`?

sihanwang41 commented on python/ray/serve/_private/constants.py:
```diff
@@ -128,6 +128,10 @@
 
+# The time in seconds that the Serve client waits before checking if
+# deployment has been created
+CLIENT_CHECK_CREATION_POLLING_INTERVAL_S: float = 0.1
```
nit: CLIENT_CHECK_DEPLOYMENT_CREATION_POLLING_INTERVAL_S
why not use CLIENT_POLLING_INTERVAL_S?

zcin replied:
CLIENT_POLLING_INTERVAL_S is 1 second, which means it will add an extra 1 second delay for `serve.run` to return. Usually a deployment should be created pretty quickly (should take one controller update iteration), so I think using a poll interval of 0.1 would be better here.

sihanwang41 commented on python/ray/serve/_private/deploy_utils.py:
```diff
@@ -97,6 +97,7 @@ def deploy_args_to_deployment_info(
     route_prefix: Optional[str],
     is_driver_deployment: Optional[bool] = False,
     app_name: Optional[str] = None,
+    **kwargs,
```
why do we need it? (not being used in this func)

zcin replied:
There are some args passed to this function from application state manager that isn't actually used by the function. Instead of having to prune the args to perfectly match the function, I added this here since it's cleaner.

sihanwang41 commented on python/ray/serve/_private/deployment_state.py:
```diff
@@ -1364,7 +1364,8 @@ def autoscale(
 
     def delete(self) -> None:
-        self._set_target_state_deleting()
+        if not self._target_state.deleting:
```
qq: any issue to fix by adding an if here?

sihanwang41 commented on python/ray/serve/controller.py:
```diff
@@ -160,7 +160,7 @@ async def __init__(
         # Manage all applications' state
         self.application_state_manager = ApplicationStateManager(
-            self.deployment_state_manager
+            self.deployment_state_manager, self.endpoint_state, self.kv_store
```
For my understanding, so the only reason to pass self.deployment_state_manager is that we have to keep the v1 api, right?

zcin replied:
Do you mean the reason why we don't construct the deployment state manager inside application state manager? If so, we would need to change the structure to be one deployment state manager per application. If we remove the v1 api, we can make that change if we want.

GeneDer commented on python/ray/serve/_private/application_state.py:
```diff
@@ -141,127 +229,191 @@ def deploy(self, deployment_params: List[Dict]) -> List[str]:
-                    num_health_deployments += 1
-
-            if len(unhealthy_deployment_names) != 0:
+        if self._check_obj_ref_ready(self._deploy_obj_ref):
```
Question: Should there be a time out for this just in case if deploy obj ref hangs forever?

zcin replied:
This function doesn't block on the object ref, it uses timeout=0. If the deploy obj ref hangs forever, _check_obj_ref_ready will continuously return false, and the application will just be stuck in DEPLOYING forever. We currently don't have anything that checks on how long an app has been deploying for - we could potentially add logic that moves it to unhealthy, but I think if the app is stuck on deploying it also is enough signal to the user that something is wrong.

GeneDer commented on python/ray/serve/_private/client.py:
```diff
@@ -219,6 +220,59 @@ def _wait_for_deployment_deleted(self, name: str, timeout_s: int = 60):
         else:
             raise TimeoutError(f"Deployment {name} wasn't deleted after {timeout_s}s.")
 
+    def _wait_for_deployment_created(self, name: str, timeout_s: int = -1):
```
Non-blocker: Would be nice to add a docstring here.

GeneDer commented on python/ray/serve/_private/client.py:
```diff
@@ -219,6 +220,59 @@ def _wait_for_deployment_deleted(self, name: str, timeout_s: int = 60):
 
+    def _wait_for_deployment_created(self, name: str, timeout_s: int = -1):
+        start = time.time()
+        while time.time() - start < timeout_s or timeout_s < 0:
```
Question: it sounds like if `timeout_s` < 0, then we will run the loop indefinitely? Given that this is not async, would it block other process?

zcin replied:
Yup it would run indefinitely. This should only be called in either the user's python script through `serve.run()`, or in the ray task kicked off separately by the controller, so it shouldn't block anything we don't intend to block.

GeneDer commented on python/ray/serve/_private/client.py:
```diff
@@ -219,6 +220,59 @@ def _wait_for_deployment_deleted(self, name: str, timeout_s: int = 60):
+        Raises TimeoutError if this doesn't happen before timeout_s.
+        """
+        start = time.time()
+        while time.time() - start < timeout_s or timeout_s < 0:
```
Same question as above, would we stuck in the loop forever and blocks other process?

GeneDer reviewed (approved):
LGTM

edoakes reviewed (commented):
See comments, I think `update` can be restructured to make the behavior more obvious from the top level. I still need to do a close review of the checkpointing/recovery logic as well.

Please also do the following:

- thoroughly manually test the failure handling behavior. Try killing the controller at various points during deploy/update/deletion and verify it works as expected.
- audit the log messages output by `application_state`. If you were debugging an issue related to a user's workload, what information would you want to see in the logs? Could you diagnose an issue purely from the statements here? Have this in mind as you manually test.

zcin commented on python/ray/serve/_private/application_state.py:
```diff
@@ -141,127 +229,191 @@ def deploy(self, deployment_params: List[Dict]) -> List[str]:
+                ApplicationStatus.DEPLOY_FAILED,
+            ]:
+                self._update_status(
+                    ApplicationStatus.DEPLOY_FAILED, status_msg=status_msg
```
This makes sense! I've refactored this function and `reconcile_deploy_obj_ref` to follow this pattern of returning the status and message, to be set at the callsite.

zcin commented:
@edoakes I've done a bunch of manual testing by killing the controller at random points, and everything I've tested works as expected! I also found the logs we currently have to be pretty useful for figuring out at what point the controller crashed, and what state was recovered during controller recovery. (I added one log for recovering config, if apps were deployed through a config)

edoakes commented on python/ray/serve/_private/application_state.py:
```diff
@@ -141,127 +237,189 @@ def deploy(self, deployment_params: List[Dict]) -> List[str]:
+        apply_deployments_args().
+        """
+
+        if self._deploy_obj_ref:
+            logger.info(
+                f'Received new config for application "{self._name}". '
+                "Cancelling previous request."
+            )
+            ray.cancel(self._deploy_obj_ref)
```
I think there is a race condition here.

Because we do this "fire and forget," there may be a (very rare) situation where we cancel this obj ref and start a new one, but this one actually ends up running and overwriting the goal state. In this case we would have an internal state inconsistency: we think the application is updated but it's actually running a previous version.

This is another reason why we should change the behavior such that the `deploy_obj_ref` _returns_ the goal state and it gets applied here: that way we can reason about and avoid any such race conditions from having multiple writers.

I don't think this is a regression from the current behavior, so I'm ok not fixing it in this PR assuming that you plan to follow up with the fix suggested above before the 2.6 branch cut.

zcin replied:
Yeah, changing the task to only return the goal state, and checking on it in application state manager will be the last part of the deploy refactor. However since I will be OOO next week, I think it's unlikely I'll be able to get this in before the 2.6 branch cut. Do you think it's acceptable to get this fix in for 2.7?

edoakes replied:
Assuming this is not a regression from existing behavior I'm OK with it going in 2.7

zcin commented:
@edoakes Tests look good, the two failing tests are failing on master. The last release test run is here: https://buildkite.com/ray-project/release-tests-pr/builds/42684#_, which also looks good. (Changes since the last run of release tests were purely log statements / adding unit tests)
